docs: update ecocredit spec (#478)
regen-network/regen-ledger: x/ecocredit/spec/01_concepts.md

# Concepts

> ## Credit Type

Author: Ok, I'll move it after credit class. Also, committed your suggestion add "main abstraction".

A credit type is the primary indicator for the methodology defined within a credit class. A credit type includes a name (e.g. carbon, biodiversity), an abbreviation (a set of 1-3 uppercase characters), a measurement unit (e.g. kilograms, tons), and a decimal precision.

The credit type abbreviation is used to construct the credit class ID. For example, "C1" is the ID for the first credit class that uses the "carbon" credit type where "C" is the credit type abbreviation and "1" is the sequence number of the credit class for the given credit type.

Credit types are listed as an on-chain parameter. Adding a new credit type to the list of approved credit types requires a parameter change proposal.

## Credit Class

> A credit class defines the class of an ecosystem service credit. A credit class includes metadata that defines a methodology for measuring and monitoring changes in ecological state. The credit class metadata defines the structure, procedures, and requirements of the methodology.
> 
> Each credit class is associated with a single credit type, which is the primary indicator for an ecosystem service methodology. Each credit class includes a credit class admin (the creator and maintainer of the credit class) and a list of approved credit issuers.

Reviewer: ```suggestion
A credit class is the primary abstraction for ecosystem service credits and is defined by 5 attributes:
- **credit class ID**: auto-generated, and auto incrementing
- **admin**: The regen address who can update fields/attributes of this credit class
- **issuer list**: The list of regen addresses who are allowed to issue credit batches under this credit class
- **credit type**: The primary indicator for this credit class (e.g. Carbon measured in Tons Equiv. CO2 sequestered)
- **metadata**: A byte array (up to 256 bytes) which can be used to store small amounts of metadata, or a URI that points to an off-chain resource for querying more complete metadata information. This usually would include descriptive information about the credit class's acceptable methodologies for monitoring changes in ecological state

Once a new credit class is created, credits can be issued at-will in distinct batches by any address in the **issuer list** of the corresponding credit class.
```

> ## Credit Class Creator

Reviewer: ```suggestion
## Credit Class Creator Allowlist
```

I think that to discuss "credit class creators" outside of the context of the allow list is a bit confusing, so maybe we should add "allowlist" here and speak about that as opposed to talking about "credit class creator" as a role in its own right?

For example, when we eventually disable the allowlist entirely, will it make much sense to talk about "credit class creator" as a distinct role, since everyone will be able to do it?

Author: Good idea! I see how it might cause confusing alongside the other two roles.

> A credit class creator is an address with the authority to create credit classes. The list of approved credit class creators is stored as an on-chain parameter that can only be updated through the governance process. Credit classes can be created  by executing a transaction on-chain with the required credit type, metadata, and list of approved issuers. Upon creation of a credit class credit, the credit class creator is set as the admin for the given credit class.

Reviewer: ```suggestion
The ecocredit module supports the option of restricting credit class creation to a permissionned set of addresses. When enabled, this list of approved credit class creators is stored as an on-chain parameter that can only be updated through the governance process. Regen Ledger 2.0 is intended to launch with this allowlist enabled.

For users wanting to experiment with creating their own credit classes on testnets, the Hambach Testnet will support permissionless credit class creation so any user can create new credit classes and test out the ecocredit module's functionality.
```

## Credit Class Admin

> The credit class admin is defined within a credit class as the address with the authority to manage and update the credit class. The credit class admin will have the ability to transfer the admin role to another address, manage the list of credit class issuers, and change credit class metadata.

Reviewer: ```suggestion
The credit class admin is defined within a credit class as the address with the authority to manage and update the credit class. When a new credit class is created, the admin is always initially set to the address that created the credit class. The credit class admin will have the ability to transfer the admin role to another address, manage the list of credit class issuers, and change credit class metadata.
```

## Credit Issuers

The credit issuers are defined within a credit class as a list of addresses with the authority to mint new credits and issue credit batches of the corresponding credit class. The list of credit issuers are defined at the time the credit class is created. The credit class admin will be able to manage the list of credit issuers for the credit class that they administer.

## Credit Batch

> Credits are issued in batches by credit issuers granted the authority to issue credits for a given credit class. A credit batch refers to a batch of credits issued at a single point in time. Each credit batch has a unique ID (i.e. denomination) that starts with the abbreviation of the credit type followed by the start date, end date, and batch sequence number.

Reviewer: Can we give an example or two of what a batch ID would look like?

Author: Yea, that would be a nice addition.

A credit batch includes information about the issuer of the credit batch and the project location, and any additional information can be attached to the metadata field. The credit batch tracks the total number of active credits and the total number of cancelled credits.

## Credits

> Credits are a loose term that refers to any fractional amount of a credit batch. Credits are issued in credit batches by an approved credit issuer within a credit class. Credit denominations are defined by the credit batch and each credit batch provides a unique denomination. Credits are either tradable or retired and each credit batch tracks the number of tradable and retired credits.

Reviewer: ```suggestion
Credits are a loose term that refers to any fractional amount of a credit batch. Credits are either tradable or retired and each credit batch tracks the number of tradable and retired credits.
```

The two middle sentences seem redundant with the above credit batch section.

## Tradable Credits

Tradable credits are credits that can be transferred by the owner to another account.

## Retired Credits

> Retired credits are credits that cannot be transferred between accounts nor can they be unretired. Retired credits are equivalent to burned tokens with the exception that retired credits are actively tracked after being retired. Retiring a credit implies that the holder of a credit is “consuming” the credit as an offset. Credits can be retired upon issuance, retired upon transfer, and retired by the owner of the credits.

Reviewer: ```suggestion
Retired credits are credits that cannot be transferred between accounts nor can they be unretired. Retired credits are equivalent to burned tokens with the exception that retired credits are actively tracked after being retired. Retiring a credit implies that the holder of a credit is “claiming” the credit as an offset. Credits can be retired upon issuance, retired upon transfer, and retired by the owner of the credits.
```

Should we mention how retirement location can be provided upon retirement?

Author: Yea, it's required upon retirement. I'll add a line to the description here.

# Concepts

## Credit Class

A credit class is the primary abstraction for ecosystem service credits and is defined by 5 attributes:
- **credit class ID**: auto-generated, and auto incrementing
- **admin**: The regen address who can update fields/attributes of this credit class
- **issuer list**: The list of regen addresses who are allowed to issue credit batches under this credit class
- **credit type**: The primary indicator for this credit class (e.g. Carbon measured in Tons Equiv. CO2 sequestered)
- **metadata**: A byte array (up to 256 bytes) which can be used to store small amounts of metadata, or a URI that points to an off-chain resource for querying more complete metadata information. This usually would include descriptive information about the credit class's acceptable methodologies for monitoring changes in ecological state.

Once a new credit class is created, credits can be issued at-will in distinct batches by any address in the **issuer list** of the corresponding credit class.

## Credit Type

> A credit type is the primary indicator for the methodology defined within a credit class. A credit type includes a name (e.g. carbon, biodiversity), an abbreviation (a set of 1-3 uppercase characters), a measurement unit (e.g. kilograms, tons), and a decimal precision.

Reviewer 2: ```suggestion
A credit type is the primary indicator used by the methodology to measure the change or impact resulting from an ecosystem service. A credit type includes a name (e.g. carbon, biodiversity), an abbreviation (a set of 1-3 uppercase characters), a measurement unit (e.g. kilograms, tons), and a decimal precision.
```
